local echo (7/7): Support simplified version of local echo

PIG208 opened the pull request:
Fixes #1441.

Stacked atop #1472.

(This branch can be used to preview the full implementation)

<details>

<summary> screenshots </summary>

| message sent | got message event |
| - | - |
| ![sending-success-before](https://github.com/user-attachments/assets/9682f14c-152c-44da-b571-cfa01e9a9849) | ![sending-success-after](https://github.com/user-attachments/assets/069cb811-d7aa-405a-97fb-9c0b173d0950) |

| message failed to send | message content tapped | recipient headers |
| - | - | - |
| ![send-failed](https://github.com/user-attachments/assets/cda06163-7ae5-42b6-b6ec-b7ef67a7b5a2) | ![sent-failed-interacted](https://github.com/user-attachments/assets/51840a33-57da-4322-8761-6ec785120eee) | ![recipient-headers](https://github.com/user-attachments/assets/987dc35a-4278-4ce6-9d34-d10510d1aacc) |

</details>

gnprice commented:
I've been running on my phone a build with this PR in it for the past few days (as described at [#mobile-team > beta release management @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/beta.20release.20management/near/2174229)). It works! Glad to see this feature nearing completion.

One observation: the loading indicator is pretty conspicuous, and feels too conspicuous in the normal case where the message promptly succeeds.

* One factor is that it causes the message to be taller, so that it and all the messages above it end up shifting down slightly after the send completes. I think this is why the design in Figma calls for the indicator to encroach on the padding instead: https://github.com/zulip/zulip-flutter/pull/1498#discussion_r2085792800

  Unlike for the editing case, though, this might be easy to fix because `OutboxMessageWithPossibleSender` is already a separate widget. If so, let's fix it.

* Even after fixing that shift, it might be helpful to de-emphasize the indicator by having it only appear after some time has elapsed, like 500ms or 1s.

PIG208 commented:
Thanks! Updated this to use 500ms, and rearranged the paddings to avoid displacement.

alya commented:
The error text reads a bit funny to me. I think "Message not sent." would be better for the first part.

chrisbobbe commented on lib/widgets/message_list.dart:
```diff
@@ -1615,22 +1616,108 @@ class OutboxMessageWithPossibleSender extends StatelessWidget {
+            padding: const EdgeInsets.symmetric(horizontal: 16),
+            child: Column(crossAxisAlignment: CrossAxisAlignment.stretch,
+              children: [
+                // This is adapated from [MessageContent].
```
nit: "adapted"

chrisbobbe reviewed (commented):
A few small comments from a very incomplete review, but wanted to post before I go out the door 🙂—

PIG208 commented:
This should address comments other than the edit-message-like change; will get to that later.

Also rebased on top of the updated "local echo (5/7)" PR, fixing some tests that relied on an earlier state machine design.

This branch has some changes to accommodate the recently added `middleMessage`/`middleItem`. The tests that fail in "msglist: Add and manage outbox message objects in message list view" will pass in the commit that follows it. I will need to double-check the changes related to `middleMessage`/`middleItem`, but they should not affect the majority of this implementation.

PIG208 commented:
I temporarily removed the last two commits because the confirmation dialog requires us to rework the implementation and tests more.

chrisbobbe reviewed (commented):
Thanks! Here's another round of review—it includes some comments I wrote yesterday on those commits that you temporarily removed in this revision, and they may or may not still apply to your new versions of those commits :)

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -1601,3 +1606,123 @@ class _RestoreEditMessageGestureDetector extends StatelessWidget {
+            backgroundColor: designVariables.foreground.withFadedAlpha(0.2)));
+
+      case OutboxMessageState.failed:
+      case OutboxMessageState.waitPeriodExpired:
```
In v0.0.29 we'd enter this state even if we got a success response from the send request, if the event didn't then show up. I suspect that's what happened in the report in #1525.

In the current revision of this PR, a success response will put the outbox-message into `waiting` state indefinitely, until the event arrives. I think showing this is OK if we've gotten neither the event nor a response from the request.

(The exact status in that case is more like "tried sending the message, it's been a while, haven't heard any response yet, so it may or may not have made it through, up to you if you want to retry". Not sure there's a good concise way to communicate that.)

PIG208 commented on lib/model/message_list.dart:
```diff
@@ -163,6 +191,7 @@ mixin _MessageSequence {
   /// Either the bottom slices of both [items] and [messages] are empty,
   /// or the first item in the bottom slice of [items] is a [MessageListMessageItem]
   /// for the first message in the bottom slice of [messages].
+  // TODO(#1453) update this in the context of outbox messages.
```
Ah, this is just a TODO for myself to update this dart doc, since we can have outbox-message related items in the list.

gnprice commented on lib/model/message_list.dart:
```diff
@@ -139,14 +156,25 @@ mixin _MessageSequence {
+  ///
+  /// Usually this should not have that many items, so we do not anticipate
+  /// performance issues with unoptimized O(N) iterations through this list.
+  final List<OutboxMessage> outboxMessages = [];
```
This adds some overhead in magnitude of O(1) (where the constant is
    the number of outbox messages in a view, expected to be small) on
    message event handling.

That is not what the word "constant" means :slightly_smiling_face: 

Instead can say:

> This adds some overhead on message-event handling, linear in the number of outbox messages in a view. We rely on that number being small.

Or could say something like "O(k), where there are k outbox messages in the view".

gnprice commented on lib/model/message_list.dart:
```diff
@@ -63,6 +64,22 @@ class MessageListMessageItem extends MessageListMessageBaseItem {
+    required super.showSender,
+    required super.isLastInBlock,
+  }) : content = ZulipContent(nodes: [
+    ParagraphNode(links: [], nodes: [TextNode(message.contentMarkdown)]),
```
```suggestion
    ParagraphNode(links: null, nodes: [TextNode(message.contentMarkdown)]),
```

See the doc on the `links` field:
```
  /// A list of all [LinkNode] descendants.
  ///
  /// An empty list is represented as null.
  /// […]
```

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -718,7 +718,11 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
           header: header,
-          trailingWhitespace: 11,
+          item: data);
+      case MessageListOutboxMessageItem():
```
This implements minimal support to display outbox message message item
    widgets in the message list, without indicators for theirs states.
    Retrieving content from failed sent requests and the full UI are
    implemented in a later commit.

Is there — ah, I guess that was in the commits you temporarily removed as mentioned at https://github.com/zulip/zulip-flutter/pull/1453#issuecomment-2903279848 .

PIG208 commented on lib/model/message_list.dart:
```diff
@@ -385,13 +451,67 @@ mixin _MessageSequence {
+  ///
+  /// This is efficient due to the expected small size of [outboxMessages].
+  void _removeOutboxMessageItems() {
+    // This loop relies on the assumption that all items that follow
+    // the last [MessageListMessageItem] are derived from outbox messages.
+    // If there is no [MessageListMessageItem] at all,
+    // this will end up removing end markers.
```
This should end up removing all items, but that's less relevant now because we used to need to restore the end markers at the end.

PIG208 commented on lib/widgets/message_list.dart:
```diff
@@ -1601,3 +1606,123 @@ class _RestoreEditMessageGestureDetector extends StatelessWidget {
+      case OutboxMessageState.waiting:
+        final designVariables = DesignVariables.of(context);
+        return Padding(
+          padding: const EdgeInsetsGeometry.only(top: 1.5, bottom: 0.5),
```
Greg mentioned that for the case of

PIG208 replied:
While we have adjusted the debounce timeout in the more recent version #1472, I think the issue with the layout shift caused by the loading indicator still exists (https://github.com/zulip/zulip-flutter/pull/1453#issuecomment-2885667930), at least for messages that don't have fancy formatting.

We have 4px total in message content's bottom padding, and the height of the progress bar is 2px. How about moving the progress bar to the top half of the bottom padding, so that there is 2px instead of 0.5px after it? Meanwhile, we can adjust the bottom padding of the edit status row to be 2px so that they are consistent.'

| 2px below | 2px below (edit-message) |
| - | - |
| ![image](https://github.com/user-attachments/assets/22a7d830-b828-49bb-8625-75b5d992d297) | ![image](https://github.com/user-attachments/assets/d82824be-52ac-4d3b-a0d0-c385ab92e22b) |

Note that the local echo indicator sits inside the bottom padding of the message content, and the one for edit-message doesn't. I think this is fine since there is label that doesn't fit within the padding for edit-message anyway.

PIG208 commented on test/model/narrow_test.dart:
```diff
@@ -220,16 +194,19 @@ void main() {
+        eg.dmOutboxMessage(from: user2, to: [user3]))).isFalse();
       check(narrow.containsMessage(
-        _TestDmMessage(allRecipientIds: [1, 2]))).isTrue();
+        eg.dmOutboxMessage(from: user1, to: [user2]))).isTrue();
```
Hmm, right. The test was not set up correctly. Adding a prep commit (ideally we would change the old helpers to use `from/to`, but I kept it minimal just so we can fix the test data), so that we can keep the refactor NFC.

PIG208 commented:
Thanks for the reviews! Pushed an update, this time with the commits I removed last time, with a bunch of prep commits. I also went through this PR's interaction with `middleMessage`/`middleItem`, and implemented the behavior discussed in a prior call with Greg, making sure that `middleItem` points to `MessageListOutboxMessageItem` in certain condition (the docs on `middleItem` has been updated to reflect the behavior I intended and implemented).

gnprice commented:
Thanks for the revision! The changes related to my review at https://github.com/zulip/zulip-flutter/pull/1453#pullrequestreview-2867241573 all look good, and the initial 5 commits all look good to me now:
c1f6b8dca msglist [nfc]: Make trailingWhitespace a constant
0b12cab63 narrow test: Make sure sender is selfUser in tests
e1fa25e8f test [nfc]: Extract {dm,stream}OutboxMessage helpers
33a2e0d04 msglist test [nfc]: Make checkInvariant compatible with MessageBase
93d89a6ac msglist [nfc]: Extract _addItemsForMessage

I haven't yet read the subsequent commits in full:
5488a7a2e msglist: Use store.senderDisplayName for sender row
d64698efa msglist [nfc]: Make _SenderRow accept MessageBase
9c0b95568 msglist: Add and manage outbox message objects in message list view
6604cf49f compose [nfc]: Make confirmation dialog message flexible
b2d6c50c2 compose test [nfc]: Move some edit message helpers out of group
cfbbdc0cc compose [nfc]: Update string to mention "edit" in name and description
c8295ae9b msglist: Adjust edit-message status row bottom padding
1fa9f24c8 msglist: Support retrieving failed outbox message content

@chrisbobbe will be taking over this PR; I'll leave those latter commits until after he's made a pass through them.

chrisbobbe commented:
Great! Merging those first five, with a small commit-message tweak:

```
$ git range-diff upstream 93d89a6a a4c564b1c
1:  c1f6b8dca = 1:  86f62c9b1 msglist [nfc]: Make trailingWhitespace a constant
2:  0b12cab63 ! 2:  75f0debbf narrow test: Make sure sender is selfUser in tests
    @@ Metadata
     Author: Zixuan James Li <zixuan@zulip.com>
     
      ## Commit message ##
    -    narrow test: Make sure sender is selfUser in tests
    +    narrow test: Make sure sender is selfUser for outbox DM messages
    +
    +    [chris: expanded commit-message summary line]
    +
    +    Co-authored-by: Chris Bobbe <cbobbe@zulip.com>
     
      ## test/model/narrow_test.dart ##
     @@ test/model/narrow_test.dart: void main() {
3:  e1fa25e8f = 3:  815b9d2f4 test [nfc]: Extract {dm,stream}OutboxMessage helpers
4:  33a2e0d04 = 4:  2829bd847 msglist test [nfc]: Make checkInvariant compatible with MessageBase
5:  93d89a6ac = 5:  a4c564b1c msglist [nfc]: Extract _addItemsForMessage
```

chrisbobbe commented:
And I've sent revised versions of the next chunk, in #1535. So that just leaves the last two commits of this PR:

c8295ae9b msglist: Adjust edit-message status row bottom padding
1fa9f24c8 msglist: Support retrieving failed outbox message content

gnprice commented on lib/widgets/message_list.dart:
```diff
@@ -1651,3 +1656,126 @@ class _RestoreEditMessageGestureDetector extends StatelessWidget {
+  Widget build(BuildContext context) {
+    final message = item.message;
+    final localMessageId = message.localMessageId;
+    final state = item.message.state;
```
Maybe just say `message.state` in the two places this is used. I feel like a bare name `state` could be confusing in the context of a widget — sounds like the state of some stateful widget.

gnprice reviewed (commented):
Thanks for picking this up! I've read through the commits after #1538:
72ffbd872 compose: Remove a redundant TypingNotifier.stoppedComposing call
3b9366e4e compose [nfc]: Remove obsoleted TODO(#720)s
d56e20fec compose [nfc]: Expand a comment to include an edge case
419e17bf5 msglist: Support retrieving failed outbox message content

except the tests of the last/main commit. All looks good except one small comment below. (And I should read the tests too.)

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -1551,6 +1553,163 @@ void main() {
+
+      await sendMessageAndFail(tester);
+      await dismissErrorDialog(tester);
+      // Navigate back to the message list page without a compose box,
+      // where the failed to send message should be visible.
+      await tester.pageBack();
+      await tester.pump(); // handle tap
+      await tester.pump(const Duration(milliseconds: 500)); // wait for navigation
```
Should this be adapted to not specify the duration, per that upstream PR?

gnprice replied:
Because customer_testing pins the version of our repo, I guess it's not an immediate problem (even for that upstream PR) if we merge it this way. But if the upstream PR lands and this is already merged, it'd break our CI.

gnprice commented on test/widgets/message_list_test.dart:
```diff
@@ -1551,6 +1553,163 @@ void main() {
+      await checkTapNotRestoreMessage(tester);
+    });
+
+    testWidgets('hidden -> failed, tap to restore message', (tester) async {
```
nit: perhaps put this above the "tapping does nothing if compose box is not offered" case — this is a lot simpler and serves as a sort of baseline for understanding the more complex cases

gnprice reviewed (commented):
Thanks! Also just read through those tests; a handful of comments below.

chrisbobbe commented on lib/widgets/compose_box.dart:
```diff
@@ -1869,6 +1872,29 @@ class _ComposeBoxState extends State<ComposeBox> with PerAccountStoreAwareStateM
+      final controller = this.controller;
+      controller
+        ..content.value = TextEditingValue(text: outboxMessage.contentMarkdown)
+        ..contentFocusNode.requestFocus();
```
This is easy to test :) done.

gnprice replied:
Cool, looks good :)

chrisbobbe commented:
Thanks for the review! Revision pushed.
